### PR TITLE
[bugfix] Resolve hanging in DDP

### LINKED ISSUE #6343: trainer.training_type_plugin.broadcast doesn't seem to work properly
### To Reproduce

Use following [**BoringModel**](https://colab.research.google.com/drive/1HvWVVTK8j2Nj52qU4Q4YCyzOm0_aLQF3?usp=sharing) and post here

### Environment

**Note**: `Bugs with code` are solved faster ! `Colab Notebook` should be made `public` !

* `IDE`: Please, use our python [bug_report_model.py](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pl_examples/bug_report_model.py
) template.

* `Colab Notebook`: Please copy and paste the output from our [environment collection script](https://raw.githubusercontent.com/PyTorchLightning/pytorch-lightning/master/tests/collect_env_details.py) (or fill out the checklist below manually).

You can get the script and run it with:
```
wget https://raw.githubusercontent.com/PyTorchLightning/pytorch-lightning/master/tests/collect_env_details.py
# For security purposes, please check the contents of collect_env_details.py before running it.
python collect_env_details.py
```

 - PyTorch Version (e.g., 1.0):
 - OS (e.g., Linux):
 - How you installed PyTorch (`conda`, `pip`, source):
 - Build command you used (if compiling from source):
 - Python version:
 - CUDA/cuDNN version:
 - GPU models and configuration:
 - Any other relevant information:

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

This PR resolves DDP hangs in ModelCheckpoint + add warning for non reduced metrics.

 @awaelchli @SeanNaren @justusschock  any idea which issues it might fix ?

Fixes #6343

## Before submitting
- [ ] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [ ] Did you make sure to update the documentation with your changes? (if necessary)
- [ ] Did you write any new necessary tests? (not for typos and docs)
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/trainer/connectors/logger_connector/epoch_result_store.py [hunk 1]
```diff
@@ -104,8 +121,25 @@ def get_batch_log_metrics(self, *args, **kwargs):
...
+
         func = getattr(opt_metric, func_name)
         metrics_to_log = func(*args, add_dataloader_idx=self.has_several_dataloaders, **kwargs)
+
+        if torch.distributed.is_initialized():
+            device = self._all_gather_fn.__self__.device
+            for non_metric_key in opt_metric.get_non_metrics_keys():
+                if non_metric_key in metrics_to_log and non_metric_key not in warning_cache.warned_metrics:
+                    metric = self._all_gather_fn(metrics_to_log[non_metric_key].to(device))
+                    if any(metric[0] != m for m in metric[1:]):
```
Comment (carmocca) [inline pytorch_lightning/trainer/connectors/logger_connector/epoch_result_store.py:133]: Can it happen that metric has only one element? I believe you can initialize distributed with one process

File: pytorch_lightning/trainer/connectors/logger_connector/epoch_result_store.py [hunk 2]
```diff
@@ -104,8 +121,25 @@ def get_batch_log_metrics(self, *args, **kwargs):
...
             raise Exception("The provided opt_metric should be a Result Object. Something is wrong")
+
         func = getattr(opt_metric, func_name)
         metrics_to_log = func(*args, add_dataloader_idx=self.has_several_dataloaders, **kwargs)
+
+        if torch.distributed.is_initialized():
+            device = self._all_gather_fn.__self__.device
+            for non_metric_key in opt_metric.get_non_metrics_keys():
+                if non_metric_key in metrics_to_log and non_metric_key not in warning_cache.warned_metrics:
+                    metric = self._all_gather_fn(metrics_to_log[non_metric_key].to(device))
```
Comment (carmocca) [inline pytorch_lightning/trainer/connectors/logger_connector/epoch_result_store.py:132]: Do we need to move to device?

File: pytorch_lightning/trainer/connectors/logger_connector/epoch_result_store.py [hunk 3]
```diff
@@ -104,8 +121,25 @@ def get_batch_log_metrics(self, *args, **kwargs):
...
+                    if any(metric[0] != m for m in metric[1:]):
+                        warning_cache.warn(
+                            f"The value associated to the key {non_metric_key}: {metric.cpu().tolist()} "
+                            "doesn't appear to be the same accross all processes. "
+                            "HINT: One could either do: self.log(..., sync_dist=True, sync_fn=torch.mean)"
+                            " to force mean reduction across processes which can be unaccurate or implement"
+                            " a ``pytorch_lightning.metrics.Metric``"
```
Comment (carmocca) [inline pytorch_lightning/trainer/connectors/logger_connector/epoch_result_store.py:139]:
```suggestion
                            "HINT: One could either do: `self.log(..., sync_dist=True, sync_fn=torch.mean)`"
                            " to force mean reduction across processes which can be unaccurate or implement"
                            " a `pytorch_lightning.metrics.Metric`"
```

File: pytorch_lightning/trainer/connectors/logger_connector/epoch_result_store.py (hunk 1 above)
Comment (justusschock) [reply pytorch_lightning/trainer/connectors/logger_connector/epoch_result_store.py:133]: Not sure this is possible, but wouldn't the list be empty in that case? And any on an empty list should be false.

File: pytorch_lightning/plugins/training_type/tpu_spawn.py [hunk 4]
```diff
@@ -201,12 +201,17 @@ def save_spawn_weights(self, model: LightningModule) -> Optional[str]:
...
     def reduce_early_stopping_decision(self, should_stop: bool) -> bool:
-        should_stop = torch.tensor(int(should_stop), device=self.lightning_module.device)
-        stop = xm.mesh_reduce('stop_signal', should_stop, sum)
-        rendezvous("pl.EarlyStoppingCallback.stop_distributed_training_check")
-        should_stop = int(stop.item()) == self.world_size
-        return should_stop
+        return self.reduce_decision(should_stop)
+
+    def reduce_model_checkpoint_decision(self, should_update_best_and_save: bool) -> bool:
+        return self.reduce_decision(should_update_best_and_save)
```
Comment (carmocca) [inline pytorch_lightning/plugins/training_type/tpu_spawn.py:214]: This is already in the parent parent, can remove.

But why don't we need all the `mesh_reduce` happiness anymore?

File: pytorch_lightning/plugins/training_type/training_type_plugin.py [hunk 5]
```diff
@@ -77,10 +77,18 @@ def barrier(self, name: Optional[str] = None) -> None:
...
+
     def reduce_early_stopping_decision(self, should_stop: bool) -> bool:
-        """Reduce the early stopping decision across all possibly spawned processes"""
+        """Reduce the early stopping decision across all processes"""
         return should_stop
 
+    def reduce_model_checkpoint_decision(self, should_update_best_and_save: bool) -> bool:
+        """Reduce the model checkpoint decision across all processes"""
+        return should_update_best_and_save
+
```
Comment (carmocca) [inline pytorch_lightning/plugins/training_type/training_type_plugin.py:91]: Is there any case where `reduce_early_stopping_decision` and `reduce_model_checkpoint_decision` reduction logic is different? Should we just have `reduce_decision`?

File: pytorch_lightning/trainer/connectors/logger_connector/epoch_result_store.py [hunk 6]
```diff
@@ -104,8 +122,24 @@ def get_batch_log_metrics(self, *args, **kwargs):
...
+        if torch.distributed.is_initialized() and get_world_size() > 1:
+            for non_metric_key in opt_metric.get_non_metrics_keys():
+                if non_metric_key in metrics_to_log and non_metric_key not in warning_cache.warned_metrics:
+                    metric = self._all_gather_fn(metrics_to_log[non_metric_key])
+                    if any(metric[0] != m for m in metric[1:]):
+                        warning_cache.warn(
+                            f"The value associated to the key {non_metric_key}: {metric.cpu().tolist()} "
+                            "doesn't appear to be the same accross all processes. "
+                            "HINT: One could either do: `self.log(..., sync_dist=True, sync_fn=torch.mean)`"
+                            " to force mean reduction across processes which can be unaccurate or implement"
```
Comment (carmocca) [inline pytorch_lightning/trainer/connectors/logger_connector/epoch_result_store.py:138]:
```suggestion
                            " to force mean reduction across processes which can be inaccurate or implement"
```

File: setup.cfg [hunk 7]
```diff
@@ -44,18 +44,15 @@ exclude_lines =
...
     pytorch_lightning/tuner/auto_gpu_select.py
     # TODO: temporary, until accelerator refactor is finished
-    pytorch_lightning/accelerators/accelerator.py
-    pytorch_lightning/plugins/training_type/*.py
-    pytorch_lightning/plugins/precision/*.py
-    pytorch_lightning/plugins/base_plugin.py
+    # pytorch_lightning/accelerators/accelerator.py
+    #pytorch_lightning/plugins/training_type/*.py
+    #pytorch_lightning/plugins/precision/*.py
+    #pytorch_lightning/plugins/base_plugin.py
```
Comment (carmocca) [inline setup.cfg:55]: Can we leave this for a different PR?

File: tests/models/test_horovod.py [hunk 8]
```diff
@@ -314,33 +334,46 @@ def _compute_batch():
...
+            return [optimizer1, optimizer2], [lr_scheduler1, lr_scheduler2]
+
+    model = TestModel()
+    model.training_epoch_end = None
+
+    num_workers = 8
+    init_lr = 0.1 * num_workers
+
+    with patch('horovod.torch.size') as mock_hvd_size:
+        mock_hvd_size.return_value = 8
```
Comment (carmocca) [inline tests/models/test_horovod.py:359]:
```suggestion
    with patch('horovod.torch.size', return_value=8) as mock_hvd_size:
```

File: tests/models/test_horovod.py [hunk 9]
```diff
@@ -314,33 +334,46 @@ def _compute_batch():
...
-#
-#     adjusted_lr1 = [pg['lr'] for pg in trainer.optimizers[0].param_groups][0]
-#     adjusted_lr2 = [pg['lr'] for pg in trainer.optimizers[1].param_groups][0]
-#
-#     # Called ones after end of epoch with gamma=0.1
-#     assert pytest.approx(init_lr * 0.1) == adjusted_lr1
-#
-#     # Called every 3 steps, meaning for 1 epoch of 11 batches, it is called 3 times with gamma=0.1
-#     assert pytest.approx(init_lr * 0.1) == adjusted_lr2
+@RunIf(skip_windows=True)
```
Comment (carmocca) [inline tests/models/test_horovod.py:337]:
```suggestion
@RunIf(skip_windows=True, horovod=True)
```

File: tests/models/test_horovod.py [hunk 10]
```diff
@@ -314,33 +334,46 @@ def _compute_batch():
...
+    with patch('horovod.torch.size') as mock_hvd_size:
+        mock_hvd_size.return_value = 8
+
+        # fit model
+        trainer = Trainer(
+            default_root_dir=tmpdir,
+            max_epochs=1,
+            limit_val_batches=0.5,
+            limit_train_batches=0.2,
+            distributed_backend='horovod'
```
Comment (carmocca) [inline tests/models/test_horovod.py:367]:
```suggestion
            accelerator='horovod'
```

File: pytorch_lightning/plugins/precision/apex_amp.py [hunk 11]
```diff
@@ -45,6 +45,8 @@ def connect(self, model: torch.nn.Module, optimizers: Sequence['Optimizer'],
             return model, optimizers, lr_schedulers
         model, optimizers = self.configure_apex(amp, model, list(optimizers), self.amp_level)
         self.reinit_scheduler_properties(optimizers, lr_schedulers)
+        for opt in optimizers:
+            opt.zero_grad()
```
[the remark this answers is not in the document]
Comment (tchaton) [reply pytorch_lightning/plugins/precision/apex_amp.py:49]: It was a wip to resolve apex + Horovod but there are downstream bugs. We decided to remove support for the combination.

[the remark this answers is not in the document]
Comment (tchaton) [reply pytorch_lightning/plugins/precision/apex_amp.py:49]: Removing it !

File: pytorch_lightning/plugins/training_type/horovod.py [hunk 12]
```diff
@@ -159,8 +159,10 @@ def reduce(self, tensor, group: Optional[Any] = None, reduce_op: Optional[Union[
         hvd.join()
         return hvd.allreduce(tensor, op=reduce_op)
 
-    def gather_all_tensors(self, result: Union[torch.Tensor], group: Optional[Any] = None):
-        if group is not None:
+    def all_gather(
+        self, result: Union[torch.Tensor], group: Optional[Any] = None, sync_grads: bool = False
+    ) -> torch.Tensor:
+        if group is not None and group != torch.distributed.group.WORLD:
```
[the remark this answers is not in the document]
Comment (tchaton) [reply pytorch_lightning/plugins/training_type/horovod.py:165]: Good call !

File: pytorch_lightning/distributed/dist.py [hunk 13]
```diff
@@ -31,32 +24,18 @@ def __init__(self, rank=None, device=None):
...
+    def broadcast(self, obj: Any, group=None):
+        is_list = isinstance(obj, list)
+
+        if not is_list:
+            obj = [obj]
+
+        if self.rank != 0:
+            obj = [None for _ in range(len(obj))]
+
+        broadcast_object_list(obj, 0, group=group or GroupMember.WORLD)
```
[the remark this answers is not in the document]
Comment (tchaton) [reply pytorch_lightning/distributed/dist.py:36]: Done !

File: pytorch_lightning/distributed/dist_utils.py [hunk 14]
```diff
@@ -0,0 +1,91 @@
+# Copyright The PyTorch Lightning team.
```
[the remark this answers is not in the document]
Comment (tchaton) [reply pytorch_lightning/distributed/dist_utils.py:1]: Done !

File: pytorch_lightning/plugins/training_type/training_type_plugin.py (hunk 5 above)
Comment (tchaton) [reply pytorch_lightning/plugins/training_type/training_type_plugin.py:91]: Removed the logic.

File: pytorch_lightning/plugins/training_type/tpu_spawn.py (hunk 4 above)
Comment (tchaton) [reply pytorch_lightning/plugins/training_type/tpu_spawn.py:214]: We need the logic to convert the boolean for decision reduction.

File: setup.cfg (hunk 7 above)
Comment (tchaton) [reply setup.cfg:55]: Yes. Sounds good to me.

File: tests/models/test_horovod.py (hunk 8 above)
Comment (tchaton) [reply tests/models/test_horovod.py:359]: Done !

File: tests/models/test_horovod.py (hunk 9 above)
Comment (tchaton) [reply tests/models/test_horovod.py]: Done !

File: tests/models/test_horovod.py (hunk 10 above)
Comment (tchaton) [reply tests/models/test_horovod.py]: Done !

File: CHANGELOG.md [hunk 15]
```diff
@@ -83,6 +83,12 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - Fixed DP reduction with collection ([#6324](https://github.com/PyTorchLightning/pytorch-lightning/pull/6324))
 
 
+- Fixed `ModelCheckpoint` file_exists using all_gather + add warnings with DDP when non-metrics aren't reduced ([#6345](https://github.com/PyTorchLightning/pytorch-lightning/pull/6345))
+
+
+- Fixed DP reduction with collection ([#6324](https://github.com/PyTorchLightning/pytorch-lightning/pull/6324))
+
+
```
Comment (carmocca) [inline CHANGELOG.md:91]: Duplicated

```suggestion
```

**Review by carmocca — APPROVED** [review]

**Review by ananthsub — COMMENTED** [review]
@tchaton are you sure its the broadcast in the checkpoint that's causing the hang? do you have a stacktrace that points to this?
